### PR TITLE
[forward ad] Fix codegen to ignore undefined outputs

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81114
* #81055
* #79996

I don't think there's a way to avoid functions returning undefined tensors as outputs, so codegen will have to detect them before calling _set_fw_grad. Alternatively, we can just make calling _set_fw_grad with undefined self a no-op, but I'm biasing toward keeping _set_fw_grad more strict in case it is called in other areas.

Fixes https://github.com/pytorch/pytorch/issues/81111

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
Sounds good in principle. Needs testing.

**Comment by soulitzer** [comment]
Testing added!

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -1419,6 +1422,7 @@ def emit_fw_derivatives() -> List[str]:
                             out_arg=res[0], is_inplace=is_inplace_str
                         )
                     )
+                    requires_fw_grad += f" && ({derivative.var_names[0]}.defined())"
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:1425]: Don't we already check above that this is defined in https://github.com/pytorch/pytorch/blob/1afb804f26c960aec32aed7c587dcb483be8c580/torch/csrc/autograd/functions/utils.h#L100 ?

**Review by albanD — APPROVED** [review]
SGTM
nit question about double defined check

File: tools/autograd/gen_variable_type.py (hunk 1 above)
Comment (soulitzer) [reply tools/autograd/gen_variable_type.py:1425]: Yeah that is only applied to the inputs of the jvp though

**Pull request closed without merging**